### PR TITLE
[#176862632] Add more specific attributes on CGN statuses

### PR DESCRIPTION
**Pull request by AleDore**
#### List of Changes
- Refactor Cgn statuses
- Refactor activities and orchestrator
- Change RowType on card expirations table, in order to avoid querying cosmos to retrieve infos about `activation_date` and `expiration_date`

#### Motivation and Context
Since a CGN could be showed in app through its statuses (except for `PENDING` status), we have to map `activation_date` and `expiration_date` in all other statuses

Please refer to related [story](https://www.pivotaltracker.com/story/show/176862632) for further informations.

#### How Has This Been Tested?
It has been tested by performing unit tests and integration through io-mock.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: StoreCgnExpirationActivity/table.ts [hunk 1]
```diff
@@ -11,14 +11,17 @@ export const insertCgnExpiration = (
...
 ): TaskEither<Error, TableService.EntityMetadata> =>
   taskify<Error, TableService.EntityMetadata>(cb =>
     tableService.insertOrReplaceEntity(
       cgnExpirationTableName,
       {
         PartitionKey: eg.String(date_fns.format(expirationDate, "yyyy-MM-dd")),
-        RowKey: eg.String(fiscalCode)
+        RowKey: eg.String(
+          JSON.stringify({ fiscalCode, activationDate, expirationDate })
+        )
```
Comment (balanza) [inline StoreCgnExpirationActivity/table.ts:24]: Not sure this is a good option, as in JS object keys order are not deterministic, hence you cannot be sure the same object is serialised into the same string (although V8 optimisation will tend to handle them alphabetically, but this cannot be trusted).

A stronger option may be an array: 
```suggestion
        RowKey: eg.String([fiscalCode, activationDate, expirationDate].join("-"))
```

Or just a concatenated string:
```suggestion
        RowKey: eg.String(`${fiscalCode}-${activationDate}-${expirationDate}`)
```

File: UpdateExpiredCgn/table.ts [hunk 2]
```diff
@@ -10,24 +12,33 @@ import {
   TableEntry
 } from "../utils/table_storage";
 
+const ExpiredCgnRowKeyType = t.interface({
```
Comment (balanza) [inline UpdateExpiredCgn/table.ts:15]: Maybe we don't need `-Type` suffix, as we already differentiate types with the initial capital letter
```suggestion
const ExpiredCgnRowKey = t.interface({
```

File: UpdateExpiredCgn/table.ts [hunk 3]
```diff
@@ -10,24 +12,33 @@ import {
...
+  fiscalCode: FiscalCode
+});
+
+export type ExpiredCgnRowKeyType = t.TypeOf<typeof ExpiredCgnRowKeyType>;
+
 /**
  * Do something with the user hash extracted from the table entry
  */
-const withFiscalCodeFromEntry = (f: (s: FiscalCode) => void) => (
+const withExpiredCgnRowFromEntry = (f: (s: ExpiredCgnRowKeyType) => void) => (
```
Comment (balanza) [inline UpdateExpiredCgn/table.ts:26]: If the above is applied
```suggestion
const withExpiredCgnRowFromEntry = (f: (s: ExpiredCgnRowKey) => void) => (
```

File: StoreCgnExpirationActivity/table.ts [hunk 4]
```diff
@@ -11,14 +11,17 @@ export const insertCgnExpiration = (
...
+  activationDate: Date,
   expirationDate: Date
 ): TaskEither<Error, TableService.EntityMetadata> =>
   taskify<Error, TableService.EntityMetadata>(cb =>
     tableService.insertOrReplaceEntity(
       cgnExpirationTableName,
       {
         PartitionKey: eg.String(date_fns.format(expirationDate, "yyyy-MM-dd")),
-        RowKey: eg.String(fiscalCode)
+        RowKey: eg.String(
```
Comment (BurnedMarshal) [inline StoreCgnExpirationActivity/table.ts:22]: Why not using `RowKey=fiscalCode` and create two props for the other values?
es:
```
ActivationDate: activationDate,
ExpirationDate: expirationDate
```
[example](https://github.com/pagopa/io-functions-backoffice/blob/master/utils/audit_logs.ts#L48)
We can use query on table storage for date/fiscalCode

File: UpdateExpiredCgn/table.ts [hunk 5]
```diff
@@ -10,24 +12,33 @@ import {
...
  * Fetches all user hashed returned by the provided paged query
  */
 export async function queryUsers(
   pagedQuery: PagedQuery
-): Promise<ReadonlySet<FiscalCode>> {
-  const entries = new Set<FiscalCode>();
-  const addToSet = withFiscalCodeFromEntry(s => entries.add(s));
+): Promise<ReadonlySet<ExpiredCgnRowKeyType>> {
+  const entries = new Set<ExpiredCgnRowKeyType>();
+  const addToSet = withExpiredCgnRowFromEntry(s => entries.add(s));
```
Comment (balanza) [inline UpdateExpiredCgn/table.ts:41]: If the above is applied:
```suggestion
): Promise<ReadonlySet<ExpiredCgnRowKey>> {
  const entries = new Set<ExpiredCgnRowKey>();
  const addToSet = withExpiredCgnRowFromEntry(s => entries.add(s));
```

File: UpdateExpiredCgn/function.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
   "bindings": [
     {
-        "schedule": "0 0 * * * *",
+        "schedule": "0 0 0 * * *",
```
Comment (BurnedMarshal) [inline UpdateExpiredCgn/function.json:4]: Why this change?

File: utils/orchestrators.ts [hunk 7]
```diff
@@ -132,9 +133,21 @@ export const terminateUpdateCgnOrchestratorTask = (
...
+        () => voidTask,
+        () =>
+          tryCatch(
+            () => client.terminate(orchestratorId, reason),
+            toError
+          ).foldTaskEither(
+            () => voidTask,
+            () => voidTask
+          )
+      )
```
Comment (balanza) [inline utils/orchestrators.ts:151]: Is all this to avoid terminate an orchestrator which is not running? What does happen if you try to terminate an orchestrator which is not running?

**Review by balanza — CHANGES_REQUESTED** [review]

File: UpsertCgnStatus/handler.ts [hunk 8]
```diff
@@ -110,7 +85,19 @@ export function UpsertCgnStatusHandler(
...
+          _.status.status !== PendingStatusEnum.PENDING
+            ? {
+                ...cgnStatus,
+                activation_date: _.status.activation_date,
+                expiration_date: _.status.expiration_date
+              }
+            : {
+                ...cgnStatus,
+                activation_date: new Date(),
+                expiration_date: new Date()
```
Comment (BurnedMarshal) [inline UpsertCgnStatus/handler.ts:98]: Why these props have the same value?

File: UpdateExpiredCgn/function.json (hunk 6 above)
Comment (BurnedMarshal) [reply UpdateExpiredCgn/function.json:4]: Ok, is executed one time every day. We are considering the [timezone](https://docs.microsoft.com/it-it/azure/azure-functions/functions-bindings-timer?tabs=csharp#ncrontab-time-zones)?

File: utils/orchestrators.ts (hunk 7 above)
Comment (AleDore) [reply utils/orchestrators.ts:151]: Yep, when trying to terminate an orchestrator which is not running, the client raises an error so we can handle it as a no effect, or even the termination has just be done :)

File: UpdateExpiredCgn/function.json (hunk 6 above)
Comment (AleDore) [reply UpdateExpiredCgn/function.json:4]: Basically the nCrontab is expressed as UTC so maybe we have to shift the schedule by 1 hour :)

File: utils/orchestrators.ts (hunk 7 above)
Comment (BurnedMarshal) [reply utils/orchestrators.ts:151]: Are we checking here a running orchestrator or an orchestrator that has a status?

File: UpsertCgnStatus/handler.ts (hunk 8 above)
Comment (AleDore) [reply UpsertCgnStatus/handler.ts:98]: You' re right. I'll correct

File: UpdateExpiredCgn/function.json (hunk 6 above)
Comment (BurnedMarshal) [reply UpdateExpiredCgn/function.json:4]: Maybe is not enough, we should consider solar/legal time or, if we haven't a _Linux Consumption plan_ we can set `WEBSITE_TIME_ZONE` into the function configuration.

File: utils/orchestrators.ts (hunk 7 above)
Comment (AleDore) [reply utils/orchestrators.ts:151]: A running orchestrator, btw if it is running it should have also a status

Comment (BurnedMarshal) [reply utils/orchestrators.ts:151]: If we want terminate only running orchestator we should use this:
```typescript
fromNullable(maybeStatus)
  .chain(
    fromPredicate(
      _ =>
        _.runtimeStatus === df.OrchestrationRuntimeStatus.Running ||
        _.runtimeStatus === df.OrchestrationRuntimeStatus.Pending
    )
)
```

File: UpdateExpiredCgn/function.json (hunk 6 above)
Comment (AleDore) [reply UpdateExpiredCgn/function.json:4]: We will provide timezone with `WEBSITE_TIME_ZONE` thanks to @uolter

File: UpdateExpiredCgn/table.ts [hunk 9]
```diff
@@ -10,24 +12,37 @@ import {
...
+const withExpiredCgnRowFromEntry = (f: (s: ExpiredCgnRowKey) => void) => (
   e: TableEntry
 ): void => {
   const rowKey = e.RowKey._;
-  return f(rowKey as FiscalCode);
+  // JSON.parse cannot throw cause we are sure rowKey is a valid Json
+  return f({
+    activationDate: e.ActivationDate._,
+    expirationDate: e.ExpirationDate._,
+    fiscalCode: rowKey
```
Comment (BurnedMarshal) [inline UpdateExpiredCgn/table.ts:34]:
```suggestion
    fiscalCode: e.RowKey._
```

**Review by BurnedMarshal — APPROVED** [review]
LGTM apart minor change
